README.md
# PHP-FPM Docker images

[![Release](https://img.shields.io/github/release/devilbox/docker-php-fpm.svg?colorB=orange)](https://github.com/devilbox/docker-php-fpm/releases)
[![](https://img.shields.io/badge/github-devilbox%2Fdocker--php--fpm-red.svg)](https://github.com/devilbox/docker-php-fpm "github.com/devilbox/docker-php-fpm")
[![lint](https://github.com/devilbox/docker-php-fpm/workflows/lint/badge.svg)](https://github.com/devilbox/docker-php-fpm/actions?workflow=lint)
[![build](https://github.com/devilbox/docker-php-fpm/workflows/build/badge.svg)](https://github.com/devilbox/docker-php-fpm/actions?workflow=build)
[![nightly](https://github.com/devilbox/docker-php-fpm/workflows/nightly/badge.svg)](https://github.com/devilbox/docker-php-fpm/actions?workflow=nightly)
[![License](https://img.shields.io/badge/license-MIT-%233DA639.svg)](https://opensource.org/licenses/MIT)

[![Discord](https://img.shields.io/discord/1051541389256704091?color=8c9eff&label=Discord&logo=discord)](https://discord.gg/2wP3V6kBj4)
[![Discourse](https://img.shields.io/discourse/https/devilbox.discourse.group/status.svg?colorB=%234CB697&label=Discourse&logo=discourse)](https://devilbox.discourse.group)

**Available Architectures:**  `amd64`, `arm64`

[![](https://img.shields.io/docker/pulls/devilbox/php-fpm.svg)](https://hub.docker.com/r/devilbox/php-fpm)

This repository will provide you fully functional PHP-FPM Docker images in different flavours,
versions and packed with different types of integrated PHP modules. It also solves the problem of **[syncronizing file permissions](doc/syncronize-file-permissions.md)** of mounted volumes between the host and the container.

:information_source: For details see **[Documentation: Syncronize File Permissions](doc/syncronize-file-permissions.md)**

This repository also allows you to quickly generate and **build your own custom PHP-FPM Docker image** with whatever PHP extension your desire for whatever PHP version you want and for any platform you're on (`amd64` or `arm64`). Jump to **[#Build your own image](#build-your-own-image)**.


| PHP-FPM          | Reference Implementation |
|:----------------:|:------------------------:|
| <a title="Docker PHP-FPM" href="https://github.com/devilbox/docker-php-fpm" ><img height="82px" src="https://raw.githubusercontent.com/devilbox/artwork/master/submissions_banner/cytopia/02/png/banner_256_trans.png" /></a> | <a title="Devilbox" href="https://github.com/cytopia/devilbox" ><img height="82px" src="https://raw.githubusercontent.com/devilbox/artwork/master/submissions_banner/cytopia/01/png/banner_256_trans.png" /></a> |
| Streamlined [PHP-FPM](https://github.com/devilbox/docker-php-fpm) images | The [Devilbox](https://github.com/cytopia/devilbox) |



<h2><img id="docker-tags" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Docker Tags</h2>

[![](https://img.shields.io/docker/pulls/devilbox/php-fpm.svg)](https://hub.docker.com/r/devilbox/php-fpm)

```bash
docker pull devilbox/php-fpm:<tag>
```

* [`5.2-base`](Dockerfiles/base/Dockerfile-5.2), [`5.3-base`](Dockerfiles/base/Dockerfile-5.3), [`5.4-base`](Dockerfiles/base/Dockerfile-5.4), [`5.5-base`](Dockerfiles/base/Dockerfile-5.5), [`5.6-base`](Dockerfiles/base/Dockerfile-5.6), [`7.0-base`](Dockerfiles/base/Dockerfile-7.0), [`7.1-base`](Dockerfiles/base/Dockerfile-7.1), [`7.2-base`](Dockerfiles/base/Dockerfile-7.2), [`7.3-base`](Dockerfiles/base/Dockerfile-7.3), [`7.4-base`](Dockerfiles/base/Dockerfile-7.4), [`8.0-base`](Dockerfiles/base/Dockerfile-8.0), [`8.1-base`](Dockerfiles/base/Dockerfile-8.1), [`8.2-base`](Dockerfiles/base/Dockerfile-8.2)
* [`5.2-mods`](Dockerfiles/mods/Dockerfile-5.2), [`5.3-mods`](Dockerfiles/mods/Dockerfile-5.3), [`5.4-mods`](Dockerfiles/mods/Dockerfile-5.4), [`5.5-mods`](Dockerfiles/mods/Dockerfile-5.5), [`5.6-mods`](Dockerfiles/mods/Dockerfile-5.6), [`7.0-mods`](Dockerfiles/mods/Dockerfile-7.0), [`7.1-mods`](Dockerfiles/mods/Dockerfile-7.1), [`7.2-mods`](Dockerfiles/mods/Dockerfile-7.2), [`7.3-mods`](Dockerfiles/mods/Dockerfile-7.3), [`7.4-mods`](Dockerfiles/mods/Dockerfile-7.4), [`8.0-mods`](Dockerfiles/mods/Dockerfile-8.0), [`8.1-mods`](Dockerfiles/mods/Dockerfile-8.1), [`8.2-mods`](Dockerfiles/mods/Dockerfile-8.2)
* [`5.2-prod`](Dockerfiles/prod/Dockerfile-5.2), [`5.3-prod`](Dockerfiles/prod/Dockerfile-5.3), [`5.4-prod`](Dockerfiles/prod/Dockerfile-5.4), [`5.5-prod`](Dockerfiles/prod/Dockerfile-5.5), [`5.6-prod`](Dockerfiles/prod/Dockerfile-5.6), [`7.0-prod`](Dockerfiles/prod/Dockerfile-7.0), [`7.1-prod`](Dockerfiles/prod/Dockerfile-7.1), [`7.2-prod`](Dockerfiles/prod/Dockerfile-7.2), [`7.3-prod`](Dockerfiles/prod/Dockerfile-7.3), [`7.4-prod`](Dockerfiles/prod/Dockerfile-7.4), [`8.0-prod`](Dockerfiles/prod/Dockerfile-8.0), [`8.1-prod`](Dockerfiles/prod/Dockerfile-8.1), [`8.2-prod`](Dockerfiles/prod/Dockerfile-8.2)
* [`5.2-slim`](Dockerfiles/slim/Dockerfile-5.2), [`5.3-slim`](Dockerfiles/slim/Dockerfile-5.3), [`5.4-slim`](Dockerfiles/slim/Dockerfile-5.4), [`5.5-slim`](Dockerfiles/slim/Dockerfile-5.5), [`5.6-slim`](Dockerfiles/slim/Dockerfile-5.6), [`7.0-slim`](Dockerfiles/slim/Dockerfile-7.0), [`7.1-slim`](Dockerfiles/slim/Dockerfile-7.1), [`7.2-slim`](Dockerfiles/slim/Dockerfile-7.2), [`7.3-slim`](Dockerfiles/slim/Dockerfile-7.3), [`7.4-slim`](Dockerfiles/slim/Dockerfile-7.4), [`8.0-slim`](Dockerfiles/slim/Dockerfile-8.0), [`8.1-slim`](Dockerfiles/slim/Dockerfile-8.1), [`8.2-slim`](Dockerfiles/slim/Dockerfile-8.2)
* [`5.2-work`](Dockerfiles/work/Dockerfile-5.2), [`5.3-work`](Dockerfiles/work/Dockerfile-5.3), [`5.4-work`](Dockerfiles/work/Dockerfile-5.4), [`5.5-work`](Dockerfiles/work/Dockerfile-5.5), [`5.6-work`](Dockerfiles/work/Dockerfile-5.6), [`7.0-work`](Dockerfiles/work/Dockerfile-7.0), [`7.1-work`](Dockerfiles/work/Dockerfile-7.1), [`7.2-work`](Dockerfiles/work/Dockerfile-7.2), [`7.3-work`](Dockerfiles/work/Dockerfile-7.3), [`7.4-work`](Dockerfiles/work/Dockerfile-7.4), [`8.0-work`](Dockerfiles/work/Dockerfile-8.0), [`8.1-work`](Dockerfiles/work/Dockerfile-8.1), [`8.2-work`](Dockerfiles/work/Dockerfile-8.2)

:information_source: For details see **[Documentation: Docker Tags](doc/docker-tags.md)**<br/>
:information_source: For details see **[Documentation: Supported Architectures](doc/supported-architectures.md)**



<h2><img id="docker-tags" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> PHP Versions</h2>

The following PHP versions are provided by this repository.

* `PHP 5.2`, `PHP 5.3`, `PHP 5.4`, `PHP 5.5`, `PHP 5.6`
* `PHP 7.0`, `PHP 7.1`, `PHP 7.2`, `PHP 7.3`, `PHP 7.4`
* `PHP 8.0`, `PHP 8.1`, `PHP 8.2`

> **Note:** Unreleased PHP versions are built from custom base images.

:information_source: For details see **[Documentation: PHP Versions](doc/php-versions.md)**<br/>
:information_source: For details see **[Documentation: Base Images](doc/base-images.md)**



<h2><img id="php-fpm-flavours" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Flavours</h2>

The provided Docker images heavily rely on inheritance to guarantee smallest possible image size. Each of them provide a working PHP-FPM server and you must decide what version works best for you. Look at the sketch below to get an overview about the two provided flavours and each of their different types.

```shell
        [PHP]            # Base FROM image (Official PHP-FPM image)
          ^              #
          |              #
          |              #
        [base]           # Streamlined base images with host user mapping
          ^              # environment variables and custom configs.
          |              #
          |              #
        [mods]           # Installs additional PHP modules
          ^              # via pecl, git and other means
          |              #
          |              #
        [prod]           # Devilbox flavour for production
          ^              # (locales, postifx, socat and injectables)
          |              # (custom *.ini files)
          |              #
        [slim]           # Devilbox flavour with only required
          ^              # cli tools to have a functional intranet.
          |              #
          |              #
        [work]           # Devilbox flavour for local development
                         # (includes backup and development tools)
                         # (sudo, custom bash and tool configs)
```

:information_source: For details see **[Documentation: Flavours](doc/flavours.md)**<br/>
:information_source: For details see **[Documentation: Base Images](doc/base-images.md)**



<h2><img id="php-extensions" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Available PHP extensions</h2>

> Click below listed extensions for details:

<!-- modules -->
[`amqp`](php_modules/amqp/)
[`apc`](php_modules/apc/)
[`apcu`](php_modules/apcu/)
[`bcmath`](php_modules/bcmath/)
[`blackfire`](php_modules/blackfire/)
[`bz2`](php_modules/bz2/)
[`calendar`](php_modules/calendar/)
[`ctype`](php_modules/ctype/)
[`curl`](php_modules/curl/)
[`date`](php_modules/date/)
[`dba`](php_modules/dba/)
[`dom`](php_modules/dom/)
[`enchant`](php_modules/enchant/)
[`ereg`](php_modules/ereg/)
[`exif`](php_modules/exif/)
[`FFI`](php_modules/ffi/)
[`fileinfo`](php_modules/fileinfo/)
[`filter`](php_modules/filter/)
[`ftp`](php_modules/ftp/)
[`gd`](php_modules/gd/)
[`gettext`](php_modules/gettext/)
[`gmp`](php_modules/gmp/)
[`hash`](php_modules/hash/)
[`iconv`](php_modules/iconv/)
[`igbinary`](php_modules/igbinary/)
[`imagick`](php_modules/imagick/)
[`imap`](php_modules/imap/)
[`interbase`](php_modules/interbase/)
[`intl`](php_modules/intl/)
[`ioncube`](php_modules/ioncube/)
[`json`](php_modules/json/)
[`ldap`](php_modules/ldap/)
[`libxml`](php_modules/libxml/)
[`lz4`](php_modules/lz4/)
[`lzf`](php_modules/lzf/)
[`mbstring`](php_modules/mbstring/)
[`mcrypt`](php_modules/mcrypt/)
[`memcache`](php_modules/memcache/)
[`memcached`](php_modules/memcached/)
[`mhash`](php_modules/mhash/)
[`mongo`](php_modules/mongo/)
[`mongodb`](php_modules/mongodb/)
[`msgpack`](php_modules/msgpack/)
[`mysql`](php_modules/mysql/)
[`mysqli`](php_modules/mysqli/)
[`mysqlnd`](php_modules/mysqlnd/)
[`OAuth`](php_modules/oauth/)
[`oci8`](php_modules/oci8/)
[`OPcache`](php_modules/opcache/)
[`openssl`](php_modules/openssl/)
[`pcntl`](php_modules/pcntl/)
[`pcre`](php_modules/pcre/)
[`PDO`](php_modules/pdo/)
[`pdo_dblib`](php_modules/pdo_dblib/)
[`PDO_Firebird`](php_modules/pdo_firebird/)
[`pdo_mysql`](php_modules/pdo_mysql/)
[`PDO_OCI`](php_modules/pdo_oci/)
[`pdo_pgsql`](php_modules/pdo_pgsql/)
[`pdo_sqlite`](php_modules/pdo_sqlite/)
[`pdo_sqlsrv`](php_modules/pdo_sqlsrv/)
[`pgsql`](php_modules/pgsql/)
[`phalcon`](php_modules/phalcon/)
[`Phar`](php_modules/phar/)
[`posix`](php_modules/posix/)
[`pspell`](php_modules/pspell/)
[`psr`](php_modules/psr/)
[`random`](php_modules/random/)
[`rdkafka`](php_modules/rdkafka/)
[`readline`](php_modules/readline/)
[`recode`](php_modules/recode/)
[`redis`](php_modules/redis/)
[`Reflection`](php_modules/reflection/)
[`session`](php_modules/session/)
[`shmop`](php_modules/shmop/)
[`SimpleXML`](php_modules/simplexml/)
[`snmp`](php_modules/snmp/)
[`soap`](php_modules/soap/)
[`sockets`](php_modules/sockets/)
[`sodium`](php_modules/sodium/)
[`solr`](php_modules/solr/)
[`SPL`](php_modules/spl/)
[`sqlite`](php_modules/sqlite/)
[`sqlite3`](php_modules/sqlite3/)
[`sqlsrv`](php_modules/sqlsrv/)
[`ssh2`](php_modules/ssh2/)
[`swoole`](php_modules/swoole/)
[`sysvmsg`](php_modules/sysvmsg/)
[`sysvsem`](php_modules/sysvsem/)
[`sysvshm`](php_modules/sysvshm/)
[`tidy`](php_modules/tidy/)
[`tokenizer`](php_modules/tokenizer/)
[`uploadprogress`](php_modules/uploadprogress/)
[`uuid`](php_modules/uuid/)
[`vips`](php_modules/vips/)
[`wddx`](php_modules/wddx/)
[`Xdebug`](php_modules/xdebug/)
[`xhprof`](php_modules/xhprof/)
[`xlswriter`](php_modules/xlswriter/)
[`xml`](php_modules/xml/)
[`xmlreader`](php_modules/xmlreader/)
[`xmlrpc`](php_modules/xmlrpc/)
[`xmlwriter`](php_modules/xmlwriter/)
[`xsl`](php_modules/xsl/)
[`yaml`](php_modules/yaml/)
[`zip`](php_modules/zip/)
[`zlib`](php_modules/zlib/)
[`zstd`](php_modules/zstd/)
<!-- /modules -->

:information_source: For details see **[Documentation: PHP Modules](doc/php-modules.md)**<br/>
:information_source: For details see **[Contributor Documentation: PHP Module definition](php_modules/README.md)**



<h2><img id="php-fpm-options" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Environment Variables</h2>

The provided Docker images offer environment variables to alter their startup behaviour.

:information_source: For details see **[Documentation: Environment Variables](doc/docker-env-variables.md)** or click on the variable name directly.

| Variable                                        | Short description                                             |
|-------------------------------------------------|---------------------------------------------------------------|
| [`DEBUG_ENTRYPOINT`][lnk_env_debug]             | Control docker log verbosity                                  |
| [`NEW_UID`][lnk_env_uid]                        | Syncronize user-id file system permissions                    |
| [`NEW_GID`][lnk_env_gid]                        | Syncronize group-id file system permissions                   |
| [`TIMEZONE`][lnk_env_timezone]                  | Set timezone                                                  |
| [`DOCKER_LOGS`][lnk_env_logs]                   | Make PHP log to file or docker logs                           |
| [`ENABLE_MODULES`][lnk_env_enable_mods]         | Enable specific PHP extensions                                |
| [`DISABLE_MODULES`][lnk_env_disable_mods]       | Disable specific PHP extensions                               |
| [`ENABLE_MAIL`][lnk_env_enable_mail]            | Control email-catch all (to not accidentally send out emails) |
| [`FORWARD_PORTS_TO_LOCALHOST`][lnk_env_forward] | Make remote ports available locally inside the container      |
| [`MYSQL_BACKUP_USER`][lnk_env_backup_user]      | Set MySQL username (for backups)                              |
| [`MYSQL_BACKUP_PASS`][lnk_env_backup_pass]      | Set MySQL password (for backups)                              |
| [`MYSQL_BACKUP_HOST`][lnk_env_backup_host]      | Set MySQL hostname (for backups)                              |

[lnk_env_debug]: doc/docker-env-variables.md#-debug_entrypoint
[lnk_env_uid]: doc/docker-env-variables.md#-new_uid
[lnk_env_gid]: doc/docker-env-variables.md#-new_gid
[lnk_env_timezone]: doc/docker-env-variables.md#-timezone
[lnk_env_logs]: doc/docker-env-variables.md#-docker_logs
[lnk_env_enable_mods]: doc/docker-env-variables.md#-enable_modules
[lnk_env_disable_mods]: doc/docker-env-variables.md#-disable_modules
[lnk_env_enable_mail]: doc/docker-env-variables.md#-enable_mail
[lnk_env_forward]: doc/docker-env-variables.md#-forward_ports_to_localhost
[lnk_env_backup_user]: doc/docker-env-variables.md#-mysql_backup_user
[lnk_env_backup_pass]: doc/docker-env-variables.md#-mysql_backup_pass
[lnk_env_backup_host]: doc/docker-env-variables.md#-mysql_backup_host



<h2><img id="php-fpm-options" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Volumes</h2>

The provided Docker images offer different volumes to be mounted.

:information_source: For details see **[Documentation: Volumes](doc/docker-volumes.md)** or click on the volume name directly.

| Volume                                             | Short description                     |
|----------------------------------------------------|---------------------------------------|
| [`/etc/php-custom.d/`][lnk_vol_php_custom]         | Add custom PHP `*.ini` files          |
| [`/etc/php-fpm-custom.d/`][lnk_vol_php_fpm_custom] | Add custom PHP-FPM `*.conf` files     |
| [`/startup.1.d/`][lnk_vol_startup1]                | Add custom startup `*.sh` files       |
| [`/startup.2.d/`][lnk_vol_startup2]                | Add custom startup `*.sh` files       |
| [`/var/log/php/`][lnk_vol_log_php]                 | Find PHP log files                    |
| [`/var/mail/`][lnk_vol_mail]                       | Find sent emails                      |
| [`/etc/supervisor/custom.d/`][lnk_vol_supervisor]  | Add custom supervisord `*.conf` files |
| [`/etc/bashrc-devilbox.d/`][lnk_vol_bashrc]        | Add custom bashrc files               |
| [`/shared/backups/`][lnk_vol_backups]              | Find MySQL backups                    |
| [`/ca/`][lnk_vol_ca]                               | Add custom Certificate Authority      |

[lnk_vol_php_custom]: doc/docker-volumes.md#-etcphp-customd
[lnk_vol_php_fpm_custom]: doc/docker-volumes.md#-etcphp-fpm-customd
[lnk_vol_startup1]: doc/docker-volumes.md#-startup1d
[lnk_vol_startup2]: doc/docker-volumes.md#-startup2d
[lnk_vol_log_php]: doc/docker-volumes.md#-varlogphp
[lnk_vol_mail]: doc/docker-volumes.md#-varmail
[lnk_vol_supervisor]: doc/docker-volumes.md#-etcsupervisorcustomd
[lnk_vol_bashrc]: doc/docker-volumes.md#-etcbashrc-devilboxd
[lnk_vol_backups]: doc/docker-volumes.md#-sharedbackups
[lnk_vol_ca]: doc/docker-volumes.md#-ca



<h2><img id="php-fpm-options" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> Ports</h2>

Have a look at the following table to see all offered exposed ports for each Docker image flavour.

<table>
 <thead>
  <tr>
   <th>Flavour</th>
   <th width="200">Port</th>
   <th>Description</th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td rowspan="1"><strong>base</strong><br/><strong>mods</strong><br/><strong>prod</strong><br/><strong>slim</strong><br/><strong>work</strong></td>
   <td><code>9000</code></td>
   <td>PHP-FPM listening port</td>
  </tr>
 </tbody>
</table>



<h2><img id="php-default-configuration" width="20" src="https://github.com/devilbox/artwork/raw/master/submissions_logo/cytopia/01/png/logo_64_trans.png"> PHP Default Configuration</h2>

Each PHP version is using the same sane default php.ini values, making it pain-free to switch versions and not having to worry about different php.ini settings.
**Note:** Flavours alway inherit the settings from its parent flavour if they have no own configuration.

| Flavour | Applied php.ini files|
|---------|------------------------------------------|

CHANGELOG.md
# Changelog


## Unreleased


## Release 0.151

### Added
- Allow to use supervisorctl to be able to reload PHP configuration without restarting


## Release 0.150

### Added
- Added `xhprof` PHP extension


## Release 0.149

### Added
- Added `net-tools` package


## Release 0.148

### Added
- Added `wscat` to work with websockets


## Release 0.147

### Fixed
- Fixed wkhtmltopdf version finder during install
- Fixed wp-cli runtime requirements (needs `less` in order to function)


## Release 0.146

This release adds lots of documentation about recently added features.

### Added
- Documentation


## Release 0.145

This is a massive restructuring release, which adds another layer on top of Ansible to easily manage/edit/add PHP tools and to configure their respective order of building and installing.

Additionally it introduces a new flavour: `slim` which is an intermediated stage between `prod` and `work`. It allows for a slim image with only required cli tools to work with the Devilbox.

### Added
- New PHP Flavour: `slim`
- Added `mhsendmail` for `arm64` architecture
- Added `wkhtmltopdf` for `arm64 architecture [#252](https://github.com/devilbox/docker-php-fpm/issues/252)
- Added `taskfile` (https://taskfile.dev/)
- Added mechanism to easily build custom images with custom set of PHP tools
- Added automated PHP tools dependency resolver (order of built is always correct)
- Added tons of documentation

### Changed
- Split out PHP tools into separate directories

### Fixed
- Fixed `xdebug` build


## Release 0.144

This is a massive restructuring release, which adds another layer on top of Ansible to easily manage/edit/add PHP extensions and to configure their respective order of building and loading.

### Added
- Added PHP extension: `lz4`
- Added PHP extension: `lzf`
- Added PHP extension: `zstd`
- Added mechanism to easily build custom images with custom set of PHP extensions
- Added automated PHP extension dependency resolver (order of built is always correct)
- Added tons of documentation
- Added Credit to contributors

### Changed
- Added serializer for Redis extension: `lz4`, `lzf` and `zstd`
- Restructured Documentation
- Split out PHP extensions into separate directories


## Release 0.143

### Added
- Added `phalcon` 5.x to PHP 8.0 and PHP 8.1


## Release 0.142

### Fixed
- Fixed `phalcon` module
- Fixed `swoole` module
- Fixed installation of wkhtmltopdf [#245](https://github.com/devilbox/docker-php-fpm/pull/245)
- FIxed installation of drupalconsole [#246](https://github.com/devilbox/docker-php-fpm/pull/246)
- Fixed installation of symfoni cli [#247](https://github.com/devilbox/docker-php-fpm/pull/247)
- Fixed installation of NodeJS
- Fixed installation of PostgreSQL client for PHP 5.6
- Fixed installation of PostgreSQL client for PHP 7.0
- Disabled Phalcon Devtools for PHP 7.4 as it breaks


## Release 0.141

### Fixed
- Fixed correct permission for `/opt/nvm` during startup


## Release 0.140

### Changed
- Ensure apt Jessie repositories are trusted beyond EOL


## Release 0.139

#### Added
- (Re-)added mongodb command line client
- (Re-)added postgresql command line client

### Changed
- Speed up `xargs` commands by using multi-CPU
- Use buildkit for building


## Release 0.138

#### Added
- Added arm64 support
- Added `vips` module for PHP 8.0
- Added `vips` module for PHP 8.1
- Added `swoole` module for PHP 8.1

#### Changed
- Separated nightly jobs


## Release 0.137

#### Fixed
- Fixed imklog: cannot open kernel log (/proc/kmsg): Operation not permitted.

#### Changed
- Ensured CI tests are platform agnostic (amd64 vs arm64)
- Ensured CI pipeline will work for long-running jobs

#### Removed
- Removed homebrew due to arm64 issues
- Removed postgres cmd client and apt repositories due to arm64 issues
- Removed mongodb cmd client and apt repositories due to arm64 issues
- Removed Ansible due to arm64 issues


## Release 0.136

#### Fixed
- Fixed `mongodb-org-shell` and `mongodb-org-tools` install

#### Added
- Re-added `mongodb` for PHP 5.3

#### Changed
- Switch PHP 5.4 base image to [devilbox/php-fpm-5.4](https://github.com/devilbox/docker-php-fpm-5.4) for potential arm64 support
- Switch PHP 5.5 base image to [devilbox/php-fpm-5.5](https://github.com/devilbox/docker-php-fpm-5.5) for potential arm64 support
- Changed base image back to Debian Jessie for PHP 5.2 and PHP 5.3


## Release 0.135

#### Fixed
- Fixed cloning of gitflow
- Fixed pidof issue on QUEMU by replacing it with pgrep [#854](https://github.com/cytopia/devilbox/issues/854)

#### Changed
- Changed PHP 5.2 and PHP 5.3 base images to Debian stretch
- Removed photoncms binaries (their GitHub organization went private)
- Removed `mongodb` extension from PHP 5.3 due to build errors
- Removed `ioncube` extension for PHP 5.2, PHP 5.3 and PHP 5.4 (arm64 only supported from PHP 5.5 onwards)
- Removed `codeception` from PHP 5.3


## Release 0.134

#### Changed
- Added extension `xdebug` to PHP 8.2


## Release 0.133

#### Added
- Added PHP 8.2: https://github.com/devilbox/docker-php-fpm-8.2


## Release 0.132

#### Fixed
- Fixed `nvm` PATH priority [#846](https://github.com/cytopia/devilbox/issues/846)

#### Added
- added extension `sqlsrv` to php 8.1
- added extension `pdo_sqlsrv` to php 8.1

#### Changed
- Changed postfix hostname to `localhost` instead of GitHub runners long name


## Release 0.131

#### Added
- Added binary `sqlite3` to all PHP images [#856](https://github.com/cytopia/devilbox/issues/856)
- Added binary `laravel` to PHP 8.0 and PHP 8.1 [#823](https://github.com/cytopia/devilbox/issues/823)
- Added AVIF support in GD for PHP 8.1 [#834](https://github.com/cytopia/devilbox/issues/834)
- Added extension `amqp` to PHP 8.0 and PHP 8.1 [#826](https://github.com/cytopia/devilbox/issues/826)
- Added extension `uploadprogress` to PHP 8.0 and PHP 8.1 [#158](https://github.com/devilbox/docker-php-fpm/pull/158)
- Added extension `imagick` to PHP 8.0 and PHP 8.1

Dockerfiles/slim/data/docker-entrypoint.sh
#!/usr/bin/env bash

set -e
set -u
set -o pipefail


###
### Globals
###

# The following global variables are available by our Dockerfile itself:
#   MY_USER
#   MY_GROUP
#   MY_UID
#   MY_GID

# Path to scripts to source
DVL_CONFIG_DIR="/docker-entrypoint.d"

# php.ini.d directory
DVL_PHP_INI_DIR="/usr/local/etc/php/conf.d"

# php-fpm conf.d directory
DVL_PHP_FPM_DIR="/usr/local/etc/php-fpm.d"

# This file holds error and access log definitions
DVL_PHP_FPM_CONF_LOGFILE="${DVL_PHP_FPM_DIR}/zzz-entrypoint-logfiles.conf"
DVL_PHP_INI_CONF_LOGFILE="${DVL_PHP_INI_DIR}/zzz-entrypoint-logfiles.ini"

# PHP-FPM log dir
DVL_FPM_LOG_DIR="/var/log/php"

# This is the log file for any mail related functions
DVL_PHP_MAIL_LOG="${DVL_FPM_LOG_DIR}/mail.log"

# Custom ini dir (to be copied to actual ini dir)
DVL_PHP_CUST_INI_DIR="/etc/php-custom.d"

# Custom PHP-FPM dir (to be copied to actual FPM conf dir)
DVL_PHP_CUST_FPM_DIR="/etc/php-fpm-custom.d"

# Supervisord config directory
DVL_SUPERVISOR_CONFD="/etc/supervisor/conf.d"


###
### Source libs
###
init="$( find "${DVL_CONFIG_DIR}" -name '*.sh' -type f | sort -u )"
for f in ${init}; do
	# shellcheck disable=SC1090
	. "${f}"
done



#############################################################
## Entry Point
#############################################################

###
### Set Debug level
###
DEBUG_LEVEL="$( env_get "DEBUG_ENTRYPOINT" "0" )"
log "info" "Debug level: ${DEBUG_LEVEL}" "${DEBUG_LEVEL}"


###
### Change uid/gid
###
set_uid "NEW_UID" "${MY_USER}"  "/home/${MY_USER}" "${DEBUG_LEVEL}"
set_gid "NEW_GID" "${MY_GROUP}" "/home/${MY_USER}" "${DEBUG_LEVEL}"


###
### Set timezone
###
set_timezone "TIMEZONE" "${DVL_PHP_INI_DIR}" "${DEBUG_LEVEL}"


###
### PHP-FPM 5.2 and PHP-FPM 5.3 Env variables fix
###
if php -v 2>/dev/null | grep -Eoq '^PHP[[:space:]]5\.(2|3)'; then
	set_env_php_fpm "/usr/local/etc/php-fpm.d/env.conf"
fi


###
### Set Logging
###
set_docker_logs \
	"DOCKER_LOGS" \
	"${DVL_FPM_LOG_DIR}" \
	"${DVL_PHP_FPM_CONF_LOGFILE}" \
	"${DVL_PHP_INI_CONF_LOGFILE}" \
	"${MY_USER}" \
	"${MY_GROUP}" \
	"${DEBUG_LEVEL}"


###
### Setup postfix
###
if is_docker_logs_enabled "DOCKER_LOGS" >/dev/null; then
	# PHP mail function should log to stderr
	set_postfix "ENABLE_MAIL" "${MY_USER}" "${MY_GROUP}" "${DVL_PHP_INI_DIR}" "/proc/self/fd/2" "1" "${DEBUG_LEVEL}"
else
	# PHP mail function should log to file
	set_postfix "ENABLE_MAIL" "${MY_USER}" "${MY_GROUP}" "${DVL_PHP_INI_DIR}" "${DVL_PHP_MAIL_LOG}" "0" "${DEBUG_LEVEL}"
fi


###
### Validate socat port forwards
###
if ! port_forward_validate "FORWARD_PORTS_TO_LOCALHOST" "${DEBUG_LEVEL}"; then
	exit 1
fi


###
### Supvervisor: supervisord.conf
###
supervisor_create_config "/etc/supervisor/supervisord.conf"


###
### Supervisor: socat
###
for line in $( port_forward_get_lines "FORWARD_PORTS_TO_LOCALHOST" ); do
	lport="$( port_forward_get_lport "${line}" )"
	rhost="$( port_forward_get_rhost "${line}" )"
	rport="$( port_forward_get_rport "${line}" )"
	supervisor_add_service \
		"socat-${lport}-${rhost}-${rport}" \
		"/usr/bin/socat tcp-listen:${lport},reuseaddr,fork tcp:${rhost}:${rport}" \
		"${DVL_SUPERVISOR_CONFD}" \
		"${DEBUG_LEVEL}"
done


###
### Supervisor: rsyslogd & postfix
###
if [ "$( env_get "ENABLE_MAIL" )" = "1" ] || [ "$( env_get "ENABLE_MAIL" )" = "2" ]; then
	supervisor_add_service "rsyslogd" "/usr/sbin/rsyslogd -n"      "${DVL_SUPERVISOR_CONFD}" "${DEBUG_LEVEL}" "1"
	supervisor_add_service "postfix"  "/usr/local/sbin/postfix.sh" "${DVL_SUPERVISOR_CONFD}" "${DEBUG_LEVEL}"
fi


###
### Supervisor: php-fpm
###
supervisor_add_service "php-fpm"  "/usr/local/sbin/php-fpm" "${DVL_SUPERVISOR_CONFD}" "${DEBUG_LEVEL}"


###
### Copy custom *.ini files
###
copy_ini_files "${DVL_PHP_CUST_INI_DIR}" "${DVL_PHP_INI_DIR}" "${DEBUG_LEVEL}"


###
### Copy custom PHP-FPM *.conf files
###
if [ "${PHP_VERSION}" = "5.2" ]; then
	copy_fpm_5_2_conf_file "${DVL_PHP_CUST_FPM_DIR}/php-fpm.xml" "${DEBUG_LEVEL}"
else
	copy_fpm_files "${DVL_PHP_CUST_FPM_DIR}" "${DVL_PHP_FPM_DIR}" "${DEBUG_LEVEL}"
fi


###
### Enable PHP Modules
###
enable_modules "ENABLE_MODULES" "${DEBUG_LEVEL}"


###
### Disable PHP Modules
###
disable_modules "DISABLE_MODULES" "${DEBUG_LEVEL}"


###
### mysqldump-secure
###
fix_mds_permissions "${MY_USER}" "${MY_GROUP}" "${DEBUG_LEVEL}"
set_mds_settings "MYSQL_BACKUP_USER" "MYSQL_BACKUP_PASS" "MYSQL_BACKUP_HOST" "${DEBUG_LEVEL}"


###
### Fix mountpoint permissions
###
if [ ! -d "/shared/backups" ]; then
	run "mkdir -p /shared/backups" "${DEBUG_LEVEL}"
fi
if [ ! -d "/shared/httpd" ]; then
	run "mkdir -p /shared/httpd" "${DEBUG_LEVEL}"
fi
run "chown ${MY_USER}:${MY_GROUP} /shared/backups" "${DEBUG_LEVEL}"
run "chown ${MY_USER}:${MY_GROUP} /shared/httpd" "${DEBUG_LEVEL}"
run "chmod 0755 /shared/backups" "${DEBUG_LEVEL}"
run "chmod 0755 /shared/httpd" "${DEBUG_LEVEL}"


###
### Update ca-certificates
###
update_ca_certificates "/ca" "${DEBUG_LEVEL}"


###
### Run custom user supplied scripts
###
execute_custom_scripts "/startup.1.d" "${DEBUG_LEVEL}"
execute_custom_scripts "/startup.2.d" "${DEBUG_LEVEL}"


###
### Startup
###
log "info" "Starting supervisord" "${DEBUG_LEVEL}"
exec "${@}"
